recipe/meta.yaml

```diff
@@ -41,7 +41,8 @@ build:
     - CUDA_CONDA_TARGET_NAME={{ cuda_target_name }}
   run_exports:
     - {{ pin_subpackage('petsc', max_pin='x.x') }}
-    - petsc * {{ scalar }}_*
+    - petsc * {{ scalar }}_*                       # [cuda_compiler_version == "None"]
+    - petsc * {{ scalar }}_cuda{{ cuda_major }}_*  # [cuda_compiler_version != "None"]
```

Comment:
Does this mean that non-cuda builds are runtime compatible with cuda builds? If no, the non-cuda builds need to specify cudanone (and the cuda builds shouldn't share a common prefix with past builds)

Reply:
Having multiple overlapping prefixes that _should_ overlap may run afoul of https://github.com/conda/conda/issues/11442 (two packages depending on different substrings are not actually installable together, despite actually being compatible)

Reply:
> Does this mean that non-cuda builds are runtime compatible with cuda builds? If no, the non-cuda builds need to specify cudanone (and the cuda builds shouldn't share a common prefix with past builds)

I thought having a higher build number for non-cuda builds would suffice.  As far as I understood, those run_exports are for build dependencies for host specified requirements.

Reply:
> Having multiple overlapping prefixes that _should_ overlap may run afoul of [conda/conda#11442](https://github.com/conda/conda/issues/11442) (two packages depending on different substrings are not actually installable together, despite actually being compatible)

Yep, that is actually what I would like to fix with this change. Isn't gonna work?

Reply:
Sorry, I'm afraid I still don't know if the cuda builds _should_ claim to be compatible with packages built against the non-cuda builds or not. That will determine the appropriate strategy.

> I thought having a higher build number for non-cuda builds would suffice. 

Higher build number only specifies _priority_ when both are otherwise equal candidates, the cuda builds are still available as candidates and other contributors may result in them being selected, for example:

- package A depends on `petsc real_*`
- package B depends on `petsc real_cuda11_*`

Installing A and B together will (try to) create an env  with `real_cuda11_` which is currently declared to be compatible with both A and B. The question is: is that right, or should that build of petsc be considered incompatible with package A?

If they should be compatible, the common prefix absolutely _should_ work, but there are problems due to what is purely a bug in conda itself: https://github.com/conda/conda/issues/11442 which I think makes this approach unavailable to us, at least until that is addressed.

Another possible avenue if they should be compatible is to leave the build string run_exports alone and add an empty metapackage that restricts petsc to cuda variants, e.g.

```
outputs:

{%- if cuda %}
- name: petsc-cuda
  requirements:
    run:
        - {{ pin_subpackage('petsc', exact=True) }}
{%- endif %}
```


where `petsc-cuda` depends exactly on the associated cuda build and is unavailable for the non-cuda builds.

`blas` variants are handled in a similar way via the `blas` metapackage. Though there it is inverted - all libblas providers depend on mutually exclusive variants of `blas` (e.g. `blas=1=openblas`). In that case, `petsc` would depend on e.g. `petsc-cuda=1=[no]cuda` and the metapackage would be defined for all variants. Advantage here is we wouldn't need to keep publishing new versions of the metapackage.

Then packages that depend on petsc with cuda should have both `petsc` and `petsc-cuda` in host dependencies.

If they should _not_ be compatible, we should change the build string prefix scheme, such that cuda builds _do not_ share a build string prefix with the existing non-cuda builds.

Reply:
@minrk @dalcinl I have tried running the petsc4py test suite using `petsc4py + petsc=*=*cuda*` from conda. There are failures (that could potentially be fixed in PETSc) indicating that, at the moment, Cuda and non-Cuda builds are incompatible. For conda, it is thus better that cuda and non-cuda builds won't share the prefix.  I can change that, but this implies we must fix petsc4py, slepc, and slepc4py feedstocks accordingly. 

given what you say

> If no, the non-cuda builds need to specify cudanone (and the cuda builds shouldn't share a common prefix with past builds)

I propose as naming convention something like `{{scalar}}_{{device}}_*` where `device = {host | cuda11 | cuda12}`. Thoughts? What is the best way of handling this with the smaller amount of changes in downstream recipes?

Reply:
> I propose as naming convention something like `{{scalar}}_{{device}}_*` where `device = {host | cuda11 | cuda12}`. Thoughts? 

Looks reasonable.

> What is the best way of handling this with the smaller amount of changes in downstream recipes?

I don't think there is much change to do, but I may be missing something. Basically, you need to add the host/cuda11/cuda12 stuff in config yaml, update the build strings, and bump build numbers.

Reply:
The run export should also be changed, right? something like

```
  run_exports:
    - {{ pin_subpackage('petsc', max_pin='x.x') }}
    - petsc * {{ scalar }}_{{ device }}*
```

Reply:
In order to avoid problems, the new cuda prefix must not overlap with the previous prefix, i.e. it must not start with `{{ scalar }}_`. Putting cuda first is one solution to this:

```yaml
- petsc * {{ scalar }}_*. # [nocuda]
- petsc * cuda{{ cuda_major }}_{{ scalar }}_*. # [cuda]
```

Reply:
So, something like this?

In petsc

```
  string: {{ scalar }}_h{{ PKG_HASH }}_{{ build }}                       # [cuda_compiler_version == "None"]
  string: cuda{{ cuda_major }}_{{ scalar }}_h{{ PKG_HASH }}_{{ build }}  # [cuda_compiler_version != "None"]
  run_exports:
    - {{ pin_subpackage('petsc', max_pin='x.x') }}
    - petsc * {{ scalar }}_*                       # [cuda_compiler_version == "None"]
    - petsc * cuda{{ cuda_major }}_{{ scalar }}_*  # [cuda_compiler_version != "None"]
```

in petsc4py

```
diff --git a/recipe/conda_build_config.yaml b/recipe/conda_build_config.yaml
index 282d42c..7cb2789 100644
--- a/recipe/conda_build_config.yaml
+++ b/recipe/conda_build_config.yaml
@@ -5,4 +5,8 @@ mpi:
 scalar:
   - real
   - complex
-  
\ No newline at end of file
+
+device:
+  - host
+  - cuda11
+  - cuda12
diff --git a/recipe/meta.yaml b/recipe/meta.yaml
index 4441431..6732737 100644
--- a/recipe/meta.yaml
+++ b/recipe/meta.yaml
@@ -1,7 +1,7 @@
 {% set name = "petsc4py" %}
 {% set version = "3.21.1" %}  # TODO: remove Cython pinning for petsc4py-3.20.0
 {% set sha256 = "ea8c6afb16541167d39f87d5fcad98c32d856fe8a2173504ef2a31c16647d53d" %}
-{% set build = 0 %}
+{% set build = 1 %}

 {% set version_xy = version.rsplit('.', 1)[0] %}
 {% if scalar == "real" %}
@@ -24,7 +24,8 @@ build:
     - export PETSC_DIR=$PREFIX
     - {{ PYTHON }} -m pip -v install --no-deps .
   skip: true  # [win]
-  string: {{ scalar }}_h{{ PKG_HASH }}_{{ build }}
+  string: {{ scalar }}_h{{ PKG_HASH }}_{{ build }}  # [device == 'host']
+  string: {{ device }}_{{ scalar }}_h{{ PKG_HASH }}_{{ build }} # [device != 'host']
   track_features:
     - petsc4py_complex  # [scalar == "complex"]

@@ -44,7 +45,8 @@ requirements:
     - pip
     - numpy
     - {{ mpi }}
-    - petsc {{ version_xy }}.* {{ scalar }}_*
+    - petsc {{ version_xy }}.* {{ scalar }}_*               # [device == 'host']
+    - petsc {{ version_xy }}.* {{ device }}_{{ scalar }}_*  # [device != 'host']
   run:
     - python
     - {{ pin_compatible('numpy') }}
```

Reply:
also related: if a package builds against a given cuda build of petsc, is the _downstream_ package necessarily confined to that version of cuda, or not? If not, they should be able to specify a cuda build without requiring a _specific_ cuda build (i.e. exclude no-cuda). If we're not sure, it's safest to be the most restrictive, and can relax restrictions later. Substring prefixes _should_ work here, but conda bugs prevent anything but the most extremely basic use of build string prefixes.

These kinds of choices, beyond a single "exactly this variant" are very hard to encode in a build string, and the feature mutex package approach may be simpler and work better, especially as we now have multiple axes to select on.

Reply:
I just pushed the `cuda{cuda_major}_` first prefix, which I believe should work.

Going foward, I think selector packages will work better, but I think this should unstick things for now.

Reply:
Yes, exactly, that `device` prefix in petsc4py should work, I think.

Reply:
Note that petsc4py doesn't really have a reason to set a custom build string. That can be left to petsc itself.

Reply:
> I just pushed the `cuda{cuda_major}_` first prefix, which I believe should work.
> 
> Going foward, I think selector packages will work better, but I think this should unstick things for now.

@minrk I force-pushed and we lost your commit since I didn't know you were modifying the recipe. check if what I did is ok

Reply:
> also related: if a package builds against a given cuda build of petsc, is the downstream package necessarily confined to that version of cuda, or not? 

They must use the same major version, that's why I have cuda11 and cuda12. So, one can use petsc_cuda12 (built with 12.4) and specify any cuda-version==12.x they want, it should work.